rpcs3/Emu/Io/RB3MidiDrums.cpp

```diff
@@ -0,0 +1,827 @@
+		}
+	}
+	//else if (bmRequestType == 0x00 && bRequest == 0x9)
+	//{
+	//  // idk what this is but if I handle it we don't get input.
+	//	rb3_midi_drums_log.success("handled -- request %x, type %x, length %x", bmRequestType, bRequest, wLength);
+	//}
+	//else if (bmRequestType == 0x80 && bRequest == 0x6)
+	//{
+	//  // idk what this is but if I handle it we don't get input.
+	//	rb3_midi_drums_log.success("handled -- request %x, type %x, length %x", bmRequestType, bRequest, wLength);
+	//}
+	else if (bmRequestType == 0x21 && bRequest == 0x9 && wLength == 8)
+	{
+		// the game uses this request to do things like set the LEDs
+		// we don't have any LEDs, so do nothing
+	}
+	else
+	{
+		rb3_midi_drums_log.error("unhandled control_transfer: request %x, type %x, length %x", bmRequestType, bRequest, wLength);
+		usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+	}
+}
+
+void usb_device_rb3_midi_drums::interrupt_transfer(u32 buf_size, u8* buf, u32 /*endpoint*/, UsbTransfer* transfer)
+{
+	transfer->fake = true;
+	transfer->expected_count = buf_size;
+	transfer->expected_result = HC_CC_NOERR;
+	// the real device takes 8ms to send a response, but there is
+	// no reason we can't make it faster
+	transfer->expected_time = get_timestamp() + 1'000;
+
+	const auto& bytes = controller::default_state;
+	if (buf_size < bytes.size())
+	{
+		rb3_midi_drums_log.warning("buffer size < %x, bailing out early (buf_size=0x%x)", bytes.size(), buf_size);
+		return;
+	}
+	memcpy(buf, bytes.data(), bytes.size());
+
+	while (true)
+	{
+		u8 midi_msg[32];
+		usz size = sizeof(midi_msg);
+
+		// This returns a double as some sort of delta time, with -1.0
+		// being used to signal an error.
+		if (rtmidi_in_get_message(midi_in, midi_msg, &size) == -1.0)
+		{
+			rb3_midi_drums_log.error("Error getting MIDI message: %s", midi_in->msg);
+			return;
+		}
+
+		if (size == 0)
+		{
+			break;
+		}
+
+		auto kit_state = parse_midi_message(midi_msg, size);
+		if (auto combo_state = combo.take_state())
+		{
+			if (combo_state->toggle_hold_kick)
+			{
+				hold_kick = !hold_kick;
+			}
+			else
+			{
+				kit_states.push_back(std::move(combo_state.value()));
+			}
+		}
+		else
+		{
+			bool is_cancel = kit_state.snare >= midi::min_velocity();
+			bool is_accept = kit_state.floor_tom >= midi::min_velocity();
+			if (hold_kick && (is_cancel || is_accept))
+			{
+				// Hold kick brings up the song category selector menu, which can be dismissed using accept/cancel buttons.
+				hold_kick = false;
+			}
+			else
+			{
+				kit_states.push_back(std::move(kit_state));
+			}
+		}
+	}
+
+	// Clean expired states.
+	auto now = std::chrono::steady_clock::now();
+	kit_states.erase(std::remove_if(std::begin(kit_states), std::end(kit_states), [&now](const rb3drums::KitState& kit_state) {
+		return now >= kit_state.expiry;
+	}), std::end(kit_states));
+
+	// Apply states to buf.
+	for (size_t i = 0; i < kit_states.size(); ++i)
```

Comment:
Either use 'usz' instead of 'size_t' or use:
`for (const rb3drums::KitState& kit_state : kit_states)`